### PR TITLE
feat: add indexer for text-format protobuf

### PR DESCRIPTION
**Pull request by justbuchanan**
Here's the initial version of the textproto indexer. I *think* it handles everything, but between the multiple versions of protobuf, I wouldn't be surprised if I'm missing some edge cases. Take a look at the test cases and let me know if you see other things I should test for.

After this PR, I'll do some refactoring work on the bazel test rules. The textproto_verifier_test.bzl added here can share a lot with the java<->proto cross-language testing rules (https://github.com/kythe/lang-proto/issues/16).

### CODE REVIEW

File: kythe/cxx/indexer/textproto/analyzer.cc [hunk 1]
```diff
@@ -0,0 +1,418 @@
...
+  void AnalyzeField(const proto::VName& file_vname, const Message& proto,
+                    const TextFormat::ParseInfoTree& parse_tree,
+                    const FieldDescriptor& field, int field_index);
+
+  proto::VName CreateAndAddAnchorNode(const proto::VName& file,
+                                      const FieldDescriptor& field,
+                                      TextFormat::ParseLocation loc);
+
+  absl::optional<proto::VName> VNameForRelPath(
+      const std::string& simplified_path) const;
```
[the remark this answers is not in the document]
Comment (justbuchanan) [reply kythe/cxx/indexer/textproto/analyzer.cc]: Unfortunately the call to `FindWithDefault()` in the implementation doesn't work with string_view, so it would require a conversion (copy) to std::string inside the method.

File: kythe/cxx/indexer/textproto/analyzer.cc [hunk 2]
```diff
@@ -0,0 +1,418 @@
...
+  return {};
+}
+
+// The TextProtoAnalyzer maintains state needed across indexing operations and
+// provides some relevant helper methods.
+class TextProtoAnalyzer {
+ public:
+  TextProtoAnalyzer(const proto::CompilationUnit& unit,
+                    absl::string_view textproto,
+                    const absl::node_hash_map<std::string, std::string>&
```
[the remark this answers is not in the document]
Comment (justbuchanan) [reply kythe/cxx/indexer/textproto/analyzer.cc]: Pointer/iterator stability shouldn't matter. Changed to flat_hash_map.

File: kythe/cxx/indexer/textproto/analyzer.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (shahms) [reply kythe/cxx/indexer/textproto/analyzer.cc]: Hrm, I'd rather avoid FindWithDefault in that case.  While your usage here is correct, it is a fundamentally unsafe API.  It's incredibly easy to construct a temporary for the "default", return a reference to that temporary and then refer to it in subsequent lines, after its lifetime has ended.

File: kythe/cxx/indexer/textproto/analyzer.cc [hunk 3]
```diff
@@ -0,0 +1,420 @@
...
+class TextprotoAnalyzer {
+ public:
+  explicit TextprotoAnalyzer(
+      const proto::CompilationUnit* unit, absl::string_view textproto,
+      const absl::flat_hash_map<std::string, std::string>*
+          file_substitution_cache,
+      KytheGraphRecorder* recorder)
+      : unit_(unit),
+        recorder_(recorder),
+        line_index_(absl::make_unique<UTF8LineIndex>(textproto)),
```
Comment (shahms) [inline kythe/cxx/indexer/textproto/analyzer.cc]: Does this need to be a unique_ptr rather than just a direct data member?

File: kythe/cxx/indexer/textproto/analyzer.cc [hunk 4]
```diff
@@ -0,0 +1,420 @@
...
+    }
+  }
+  return absl::nullopt;
+}
+
+// The TextprotoAnalyzer maintains state needed across indexing operations and
+// provides some relevant helper methods.
+class TextprotoAnalyzer {
+ public:
+  explicit TextprotoAnalyzer(
```
Comment (shahms) [inline kythe/cxx/indexer/textproto/analyzer.cc]: minor: It seems likely that this class should be non-copyable/movable.

File: kythe/cxx/indexer/textproto/analyzer.cc [hunk 5]
```diff
@@ -0,0 +1,420 @@
...
+      const proto::CompilationUnit* unit, absl::string_view textproto,
+      const absl::flat_hash_map<std::string, std::string>*
```
Comment (shahms) [inline kythe/cxx/indexer/textproto/analyzer.cc]: Add a doc comment to the constructor stating that the pointer arguments must outlive TextprotoAnalyzer instance?

File: kythe/cxx/indexer/textproto/analyzer.cc [hunk 6]
```diff
@@ -0,0 +1,420 @@
...
+  // 1-indexed line numbers.
+  loc.line++;
+
+  if (loc.line == 0) {
+    // When AnalyzeField() is called for repeated fields or extensions, we know
+    // the field was actually present in the input textproto. Failure to find
+    // its location is an error. For other "normal" fields, failure to find a
+    // location just indicates that the field isn't present.
+    if (field_index != kNonRepeatedFieldIndex || field.is_extension()) {
+      LOG(FATAL) << "Failed to find location of field: " << field.full_name()
```
Comment (shahms) [inline kythe/cxx/indexer/textproto/analyzer.cc]: DFATAL?

File: kythe/cxx/indexer/textproto/analyzer.cc [hunk 7]
```diff
@@ -0,0 +1,420 @@
...
+    if (args->at(i) == "--proto_message") {
+      if (i + 1 < args->size()) {
+        std::string v = args->at(i + 1);
+        args->erase(args->begin() + i, args->begin() + i + 2);
+        return v;
+      }
+      return {};
+    }
+  }
+  return {};
```
Comment (shahms) [inline kythe/cxx/indexer/textproto/analyzer.cc]: absl::nullopt for these

File: kythe/cxx/indexer/textproto/analyzer.h [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+#include "kythe/cxx/common/file_vname_generator.h"
+#include "kythe/cxx/common/indexing/KytheGraphRecorder.h"
+#include "kythe/cxx/common/status.h"
+#include "kythe/proto/analysis.pb.h"
+
+namespace kythe {
+namespace lang_textproto {
+
+// The canonical name for the textproto language in Kythe.
+static const char kLanguageName[] = "protobuf_textformat";
```
Comment (shahms) [inline kythe/cxx/indexer/textproto/analyzer.h]: Sadly, until C++17 gives us inline constexpr variables, this is not safe in a header.  Prefer one of:

extern const char kLanguageName[];
extern const absl::string_view kLanguageName;
constexpr absl::string_view LanguageName() { return "protobuf_textformat";}

File: kythe/cxx/indexer/textproto/textproto_indexer_main.cc [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+    DecodeKzipFile(FLAGS_index_file,
+                   [&](const proto::CompilationUnit& unit,
+                       std::vector<proto::FileData> file_data) {
+                     Status status = lang_textproto::AnalyzeCompilationUnit(
+                         unit, file_data, &recorder);
+                     CHECK(status.ok()) << status;
+                   });
+  }
+
+  CHECK(::close(write_fd) == 0) << "Error closing output file";
```
Comment (shahms) [inline kythe/cxx/indexer/textproto/textproto_indexer_main.cc]: minor: just `raw_output.SetCloseOnDelete(write_fd)` if you're not doing anything meaningful on a close failure.

File: kythe/cxx/indexer/textproto/analyzer.cc (hunk 3 above)
Comment (justbuchanan) [reply kythe/cxx/indexer/textproto/analyzer.cc]: Good point, I'm not sure how it ended up as a pointer.
